src/liballoc/collections/vec_deque.rs

```diff
@@ -2177,8 +2177,28 @@ impl<'a, T> Iterator for Iter<'a, T> {
         Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
     {
         let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
-        let accum = front.iter().try_fold(init, &mut f)?;
-        back.iter().try_fold(accum, &mut f)
+        let mut accum = init;
+        for slice in &[front, back] {
+            let mut iter = slice.iter();
+            let res = iter.try_fold(accum, f);
+            self.tail = unsafe { self.ring.offset(iter.as_slice().as_ptr()) };
+            accum = res?;
+        }
+        accum
+    }
+
+    fn try_rfold<B, F, R>(&mut self, init: B, mut f: F) -> R where
+        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    {
+        let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
+        let mut accum = init;
+        for slice in &[back, front] {
+            let mut iter = slice.iter();
+            let res = iter.try_rfold(accum, f);
+            self.tail = unsafe { self.ring.offset(iter.as_slice().as_ptr()) };
```

Comment:
A comment justifying this would be good.

Reply:
Same here.